File: genetisk/src/main/java/no/asmund/genetisk/Creature.java
package no.asmund.genetisk;

import javafx.scene.paint.Color;

import no.asmund.genetisk.Obstacle;

import static java.lang.Math.sqrt;

public class Creature {
    float x, y;
    float startX = 400;
    float startY = 300;
    float angle;
    DNA dna;
    int geneCounter;
    Color color;
    float recordDist;
    float fitness;
    int d; //diameter
    boolean completed;

    public Creature(DNA dna) {
        this.dna = dna;
        geneCounter = 0;
        x = startX;
        y = startY;

        d = 5;
        float r = ((dna.genes[10] * dna.genes[10]) / 100);
        float g = ((dna.genes[20] * dna.genes[20]) / 100);
        float b = ((dna.genes[30] * dna.genes[30]) / 100);
        color = new Color(r, g, b, 1);
        angle = (float) (dna.genes[10] * dna.genes[10] * 3.6);
    }

    void run() {
        if (!completed) {
            update();
            geneCounter++;
        } else {
            System.out.println("ferdig");
        }

    }

    void fitness() {
        float x2 = StartApp.targetX;
        float y2 = StartApp.targetY;
        recordDist = (float) sqrt((x - x2) * (x - x2) + (y - y2) * (y - y2));
        fitness = 1 / recordDist;
        if (recordDist < 50) {
            completed = true;
            fitness *= 10;
        }
        geneCounter = 0;
    }

    void update() {
        float dx = (float) ((Math.cos(Math.toRadians(angle)) * 2));
        float dy = (float) ((Math.sin(Math.toRadians(angle)) * 2));

        // Bounce when hitting canvas edges
        if (x + dx < 0 || x + dx > StartApp.WIDTH) {
            angle = 180 - angle;
            dx = -dx;
        }
        if (y + dy < 0 || y + dy > StartApp.HEIGHT) {
            angle = -angle;
            dy = -dy;
        }

        // Test collision with obstacles
        for (Obstacle o : StartApp.obstacles) {
            if (o.collides(x + dx, y + dy)) {
                angle = angle + 180;
                dx = -dx;
                dy = -dy;
                break;
            }
        }

        x += dx;
        y += dy;
        angle += dna.genes[geneCounter];
    }
}

File: genetisk/src/main/java/no/asmund/genetisk/Creature.kt
package no.asmund.genetisk

import javafx.scene.paint.Color
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Represents a creature in the genetic algorithm simulation.
 * Each creature has DNA that determines its behavior and appearance.
 */
class Creature(val dna: DNA) {
    var x: Float = 400f  // Starting X position
    var y: Float = 300f  // Starting Y position
    val startX: Float = 400f
    val startY: Float = 300f
    var angle: Float
    var geneCounter: Int = 0
    val color: Color
    var recordDist: Float = 0f
    var fitness: Float = 0f
    val d: Int = 5  // Diameter
    var completed: Boolean = false

    init {
        // Calculate color based on genes
        val r = (dna.genes[10] * dna.genes[10]) / 100
        val g = (dna.genes[20] * dna.genes[20]) / 100
        val b = (dna.genes[30] * dna.genes[30]) / 100
        color = Color(r.toDouble(), g.toDouble(), b.toDouble(), 1.0)
        
        // Calculate initial angle based on genes
        angle = (dna.genes[10] * dna.genes[10] * 3.6).toFloat()
    }

    /**
     * Updates the creature's position and orientation based on its DNA.
     */
    fun run() {
        if (!completed) {
            update()
            geneCounter++
        } else {
            println("completed")
        }
    }

    /**
     * Calculates the fitness of the creature based on its distance to the target.
     * Closer creatures have higher fitness.
     */
    fun fitness() {
        val x2 = StartApp.targetX.toFloat()
        val y2 = StartApp.targetY.toFloat()
        
        // Calculate distance to target
        recordDist = sqrt((x - x2) * (x - x2) + (y - y2) * (y - y2))
        
        // Fitness is inverse of distance (closer = higher fitness)
        fitness = 1 / recordDist
        
        // Bonus for reaching the target
        if (recordDist < 50) {
            completed = true
            fitness *= 10f
        }
        
        geneCounter = 0
    }

    /**
     * Updates the creature's position and angle based on its DNA.
     */
    fun update() {
        // Move in the current direction
        x += (cos(Math.toRadians(angle.toDouble())) * 2).toFloat()
        y += (sin(Math.toRadians(angle.toDouble())) * 2).toFloat()
        
        // Update angle based on current gene
        angle += dna.genes[geneCounter]
    }
}